src/engine/TransitivePath.cpp

```diff
@@ -658,119 +449,158 @@ void TransitivePath::computeTransitivePathRightBound(
-    shared_ptr<const ResultTable> rightRes = _rightSideTree->getResult();
-    size_t rightWidth = rightRes->idTable().numColumns();
-    CALL_FIXED_SIZE((std::array{subWidth, rightWidth, _resultWidth}),
-                    &TransitivePath::computeTransitivePathRightBound, this,
-                    &idTable, subRes->idTable(), rightRes->idTable(),
-                    _rightSideCol, _leftIsVar, _leftSubCol, _rightSubCol,
-                    _leftValue, _minDist, _maxDist, _resultWidth);
-  } else {
-    CALL_FIXED_SIZE(subWidth, &TransitivePath::computeTransitivePath, this,
-                    &idTable, subRes->idTable(), _leftIsVar, _rightIsVar,
-                    _leftSubCol, _rightSubCol, _leftValue, _rightValue,
-                    _minDist, _maxDist);
+      rowIndex++;
+    }
   }
-
-  LOG(DEBUG) << "TransitivePath result computation done." << std::endl;
-  // NOTE: The only place, where the input to a transitive path operation is not
-  // an index scan (which has an empty local vocabulary by default) is the
-  // `LocalVocabTest`. But it doesn't harm to propagate the local vocab here
-  // either.
-  return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
 }
 
 // _____________________________________________________________________________
-std::shared_ptr<TransitivePath> TransitivePath::bindLeftSide(
-    std::shared_ptr<QueryExecutionTree> leftop, size_t inputCol) const {
-  return bindLeftOrRightSide(std::move(leftop), inputCol, true);
+template <size_t WIDTH>
+void TransitivePath::fillTableWithHull(IdTableStatic<WIDTH>& table, Map hull,
+                                       size_t startSideCol,
+                                       size_t targetSideCol) {
+  size_t rowIndex = 0;
+  for (auto const& [node, linkedNodes] : hull) {
+    for (Id linkedNode : *linkedNodes) {
+      table.emplace_back();
+      table(rowIndex, startSideCol) = node;
+      table(rowIndex, targetSideCol) = linkedNode;
+
+      rowIndex++;
+    }
+  }
 }
 
 // _____________________________________________________________________________
-std::shared_ptr<TransitivePath> TransitivePath::bindRightSide(
-    std::shared_ptr<QueryExecutionTree> rightop, size_t inputCol) const {
-  return bindLeftOrRightSide(std::move(rightop), inputCol, false);
+template <size_t SUB_WIDTH, size_t SIDE_WIDTH>
+std::pair<TransitivePath::Map, std::vector<Id>>
+TransitivePath::setupMapAndNodes(const IdTable& sub,
+                                 const TransitivePathSide& startSide,
+                                 const TransitivePathSide& targetSide,
+                                 const IdTable& startSideTable) const {
+  std::vector<Id> nodes;
+  Map edges = setupEdgesMap<SUB_WIDTH>(sub, startSide, targetSide);
+
+  // Bound -> var|id
+  nodes = setupNodesVector<SIDE_WIDTH>(startSideTable,
+                                       startSide.treeAndCol.value().second);
+
+  return std::make_pair(edges, nodes);
```

Comment:
You should definitely `std::move` the results. 
And theres no need to delare the `nodes` variable upfront.
And you can since C++17 sim0ply use the `pair` constructor, so just
` return {std::move(edges) std::move(nodes)}`.